pkg/highlight/highlighter.go

```diff
@@ -73,135 +73,180 @@ func findIndex(regex *regexp.Regexp, skip *regexp.Regexp, str []byte) []int {
+	if start <= end && end <= len(fullHighlights) {
+		for i := start; i < end; i++ {
+			fullHighlights[i] = group
 		}
 	}
+}
 
-	var nestedRegion *region
-	nestedLoc := []int{lineLen, 0}
-	searchNesting := true
-	endLoc := findIndex(curRegion.end, curRegion.skip, line)
-	if endLoc != nil {
-		if start == endLoc[0] {
-			searchNesting = false
-		} else {
-			nestedLoc = endLoc
-		}
+func (h *Highlighter) highlightPatterns(fullHighlights []Group, start int, lineNum int, line []byte, curRegion *region) {
+	lineLen := util.CharacterCount(line)
+	if lineLen == 0 {
+		return
 	}
-	if searchNesting {
-		for _, r := range curRegion.rules.regions {
-			loc := findIndex(r.start, r.skip, line)
-			if loc != nil {
-				if loc[0] < nestedLoc[0] {
-					nestedLoc = loc
-					nestedRegion = r
-				}
+
+	var patterns []*pattern
+	if curRegion == nil {
+		patterns = h.Def.rules.patterns
+	} else {
+		patterns = curRegion.rules.patterns
+	}
+
+	for _, p := range patterns {
+		if curRegion == nil || curRegion.group == curRegion.limitGroup || p.group == curRegion.limitGroup {
+			matches := findAllIndex(p.regex, nil, line)
+			for _, m := range matches {
+				h.highlightRange(fullHighlights, start+m[0], start+m[1], p.group)
 			}
 		}
 	}
-	if nestedRegion != nil && nestedLoc[0] != lineLen {
-		highlights[start+nestedLoc[0]] = nestedRegion.limitGroup
-		slice := util.SliceEnd(line, nestedLoc[1])
-		h.highlightEmptyRegion(highlights, start+nestedLoc[1], lineNum, slice)
-		h.highlightRegion(highlights, start+nestedLoc[1], lineNum, slice, nestedRegion)
-		return highlights
-	}
+}
 
-	fullHighlights := make([]Group, lineLen)
-	for i := 0; i < len(fullHighlights); i++ {
-		fullHighlights[i] = curRegion.group
+func (h *Highlighter) highlightRegions(fullHighlights []Group, start int, lineNum int, line []byte, curRegion *region, regions []*region, nestedRegion bool) {
+	lineLen := util.CharacterCount(line)
+	if lineLen == 0 {
+		return
 	}
 
-	if searchNesting {
-		for _, p := range curRegion.rules.patterns {
-			if curRegion.group == curRegion.limitGroup || p.group == curRegion.limitGroup {
-				matches := findAllIndex(p.regex, line)
-				for _, m := range matches {
-					if endLoc == nil || m[0] < endLoc[0] {
-						for i := m[0]; i < m[1]; i++ {
-							fullHighlights[i] = p.group
+	h.highlightPatterns(fullHighlights, start, lineNum, line, curRegion)
+
+	lastStart := 0
+	regionStart := 0
+	regionEnd := 0
+	for _, r := range regions {
+		if !nestedRegion && curRegion != nil && (curRegion.group != r.group || curRegion.start != r.start) {
+			continue
+		}
+		startMatches := findAllIndex(r.start, r.skip, line)
+		endMatches := findAllIndex(r.end, r.skip, line)
+		samePattern := false
+	startLoop:
+		for startIdx := 0; startIdx < len(startMatches); startIdx++ {
+			if startMatches[startIdx][0] < lineLen {
+				for endIdx := 0; endIdx < len(endMatches); endIdx++ {
+					if startMatches[startIdx][0] == endMatches[endIdx][0] {
+						// start and end are the same (pattern)
```

Comment:
What do you mean by "the same"? `r.start` and `r.end` may be different patterns, right?

...Every single line in this `if startMatches[startIdx][0] == endMatches[endIdx][0]` branch, including the comments, looks completely confusing to me. Could you explain what is it actually doing? What is the meaning of `samePattern`, why are we comparing lengths of `startMatches` and `endMatches`, why are we comparing `curRegion` and `r`, and so on.

(Maybe other parts of `storeRegions()` are no less confusing, but I haven't gotten to them yet.)

Reply:
> What do you mean by "the same"? `r.start` and `r.end` may be different patterns, right?

Right, they may be different, but are often the same as given as example in the commit message "highlighter: Rework of matching approach to speed up the processing" of the first big rework commit.
Often we've to deal with regions defined by the usual string markers/patterns (`"I'm a string"`), in which the `start` is the same as the `end` of this region...it's both time `"`.
Having the same `start` and `end` marker/pattern makes it very hard to identify where something starts and where it ends, especially when the region goes across different lines (where a next region of the same type could start too):

e.g.:
```
Here starts a "I'm a multi-line string,
which ends here" and "that's a second
one here".
```
Look at the nested `'`, which can be again a further (multi-line) region with the same `start` and `end` marker of `'` in the same syntax definition...as we've a lot of them. This creates overlapping regions.

This is the reason why we've to do some kind of "smart guessing". These guesses are, at least from my perspective commented (yes, I'm biased, since it's my code :stuck_out_tongue_winking_eye:) and even prepared with `log.Println()` to help understand, why we've to check lengths for overlapping regions.
Every region of the same type can be identified by `highlightRegions()` for a overlap situation. Since different types of regions are handled independent of each other it needs a bit more complex rules to find out if such regions overlap, which is were `storeRange()` comes in.

I fully agree, that the code isn't really trivial or easy to understand, but I wasn't able to find a better approach so far.

Reply:
> Right, they may be different, but are often the same

So it is not true that they are the same, what is true is that they _may be_ the same? Or maybe it is not important? Anyway, the comment is, at least, misleading?

It is still very unclear to me what is this `if startMatch[0] == endMatch[0]` branch doing, and I still got almost no answers to my questions (why are we comparing the _number_ of start matches with the _number_ of end matches, why are we comparing `curRegion` and `r` etc), but, per my comment https://github.com/zyedidia/micro/pull/3127#discussion_r1898727290, I hope we can rework the code so that this logic goes away and these questions become irrelevant...